torch/csrc/jit/codegen/cuda/transform_replay.cpp

```diff
@@ -644,24 +644,173 @@ std::pair<TensorDomain*, unsigned int> TransformReplay::replayCasP(
   return replayCasP(consumer, producer, compute_at_axis, root_map);
 }
 
+namespace {
+
+int getMatchedLeafPosWithoutReplay(
+    const TensorView* producer,
+    const TensorView* consumer,
+    int consumer_or_producer_pos,
+    bool consumer_pos = true) {
+  FUSER_PERF_SCOPE("transform_replay.cpp::getMatchedLeafPosWithoutReplay");
+
+  const auto c2p_root_map =
+      PairwiseRootDomainMap(producer, consumer)
+          .mapConsumerToProducer(consumer->domain(), producer->domain());
+
+  // IterDomains in consumer root also in producer root
+  std::unordered_set<Val*> mapped_consumer_roots;
+  for (auto entry : c2p_root_map) {
+    mapped_consumer_roots.emplace(entry.first);
+  }
+
+  const auto consumer_domain = consumer->domain()->domain();
+
+  auto mapped_consumer_domain_ids_vec = DependencyCheck::getAllValsBetween(
+      mapped_consumer_roots, {consumer_domain.begin(), consumer_domain.end()});
+
+  std::unordered_set<Val*> mapped_consumer_domain_ids(
+      mapped_consumer_domain_ids_vec.begin(),
+      mapped_consumer_domain_ids_vec.end());
+
+  const auto producer_domain = producer->domain()->domain();
+
+  auto it_consumer = consumer_domain.begin();
+  auto it_producer = producer_domain.begin();
+
+  auto best_effort_PasC = BestEffortReplay::replayPasC(
+      producer, consumer, -1, PairwiseRootDomainMap(producer, consumer));
+
+  auto c2p_map = best_effort_PasC.getReplay();
+
+  int mismatched_consumer_pos = 0;
+  int mismatched_producer_pos = 0;
+  while (it_consumer != consumer_domain.end()) {
+    auto consumer_id = *it_consumer;
+    if (!mapped_consumer_domain_ids.count(consumer_id)) {
+      ++it_consumer;
+      mismatched_consumer_pos++;
+      continue;
+    }
+
+    auto c2p_it = c2p_map.find(consumer_id);
+    if (c2p_it == c2p_map.end()) {
+      break;
+    }
+
+    if (it_producer == producer_domain.end()) {
+      break;
+    }
+
+    auto producer_id = *it_producer;
+
+    if (c2p_it->second == producer_id) {
+      ++mismatched_consumer_pos;
+      ++mismatched_producer_pos;
+      ++it_consumer;
+      ++it_producer;
+      if (consumer_pos) {
+        if (consumer_or_producer_pos == mismatched_consumer_pos) {
```

Comment:
Yeah, this was what was confusing me. We're not returning the actual mismatched location, we're just returning the provided position if the mismatched position is further to the right. So this function can only return the provided position, or -1 meaning there's a mismatch before we hit the provided position.

Reply:
We should either return this function to just being a bool, or we should actually return the position of the mismatch and compare that position outside of this function.

Reply:
> we're just returning the provided position if the mismatched position is further to the right.

I think we are not just returning the provided position, but instead, if you provide a consumer position, we are returning the corresponding producer position of the given consumer position. In this case, we can not return the actual mismatch position, because it does not provide the information about the corresponding producer position of the given consumer position?

I copy-pasted this function from computeAt, and actually, there is one thing that I don't understand about this function. Why are we skipping unmappable dims in the consumer but not in the producer? What breaks the symmetry?

Reply:
> I copy-pasted this function from computeAt

Yeah it's my fault, this function is really just returning yes or no, it shouldn't be returning an int. It could return an int, but should be rewritten to do so.

> Why are we skipping unmappable dims in the consumer but not in the producer? What breaks the symmetry?
Unmappable dims (might be a bad name) come from patterns associated with reductions, and really associated with normalization patterns.

If I have:
T1 = set(T0)
T2 = sum(T1, {1})
T3 = broadcast(T2, {false, true})
T4 = add(T1, T3)

T1 cannot be fully inlined to T4. So T1's second dimension (I believe) is marked as unmappable to T4. It's not okay to inline T2 into T4, but it's fine to inline T1 with T2.

Reply:
@naoyam can tell me if I'm wrong on the details of the example, but I believe the principles stand.

Reply:
> it shouldn't be returning an int. It could return an int, but should be rewritten to do so.

I still don't understand. Are you saying that the algorithm we used to compute the corresponding position is wrong, and we should rewrite to compute it differently? Or are you saying that the algorithm is OK, but we should split up the testing of "needs play" and "find the corresponding position" into two things? Or something else?

> If I have:
T1 = set(T0)
T2 = sum(T1, {1})
T3 = broadcast(T2, {false, true})
T4 = add(T1, T3)

Hmm, I am not sure if this example is a related. I think `getMatchedLeafPosWithoutReplay` only looks at `PairwiseRootDomainMap`, which will map all the dims in T4 to T1?
```C++
TEST_F(NVFuserTest, TMP) {
  auto fusion = std::make_unique<Fusion>();
  FusionGuard fg(fusion.get());

  auto tv0 = makeSymbolicTensor(2);
  fusion->addInput(tv0);
  auto tv1 = set(tv0);
  auto tv2 = sum(tv1, {1});
  auto tv3 = broadcast(tv2, {false, true});
  auto tv4 = add(tv1, tv3);
  fusion->addOutput(tv4);

  auto producer = tv1;
  auto consumer = tv4;

  const auto c2p_root_map =
      PairwiseRootDomainMap(producer, consumer)
          .mapConsumerToProducer(consumer->domain(), producer->domain());

  // IterDomains in consumer root also in producer root
  std::unordered_set<Val*> mapped_consumer_roots;
  for (auto entry : c2p_root_map) {
    mapped_consumer_roots.emplace(entry.first);
  }

  fusion->print();
  std::cout << ir_utils::toString(mapped_consumer_roots) << std::endl;
}
```
This gives `iS8{i1}, iS9{i2}`.

I think "unmapped IDs in consumer but not in producer" here refers to new broadcasting dims, and we are saying that, if we see a leaf ID that completely comes from new broadcasting dims, then we can ignore it. But why don't we want to symmetrically do the same for reductions in the producer? For non-trivial reduction, I think this is because it could not be inlined, but for trivial reduction, it should be safe to just ignore it as well?

Reply:
I believe the algorithm is correct, and it should not be symmetric. I was considering trivial reduction, but trivial reduction could likely be ignored as well.

I was trying to say the asymmetry is due to the unmappable dims, which are related to particular graph patterns with reduction (the pattern mentioned above).

Reply:
This is my mistake, I misread the return statements, I thought we were checking the return value was equal to the provided value to return that value. We're swapping produce/consume so they're different.

Reply:
The algorithm is actually not correct😋: https://github.com/csarofeen/pytorch/pull/1791

Reply:
> > it shouldn't be returning an int. It could return an int, but should be rewritten to do so.
> 
> I still don't understand. Are you saying that the algorithm we used to compute the corresponding position is wrong, and we should rewrite to compute it differently? Or are you saying that the algorithm is OK, but we should split up the testing of "needs play" and "find the corresponding position" into two things? Or something else?
> 
> > If I have:
> > T1 = set(T0)
> > T2 = sum(T1, {1})
> > T3 = broadcast(T2, {false, true})
> > T4 = add(T1, T3)
> 
> Hmm, I am not sure if this example is a related. I think `getMatchedLeafPosWithoutReplay` only looks at `PairwiseRootDomainMap`, which will map all the dims in T4 to T1?
> 
> ```c++
> TEST_F(NVFuserTest, TMP) {
>   auto fusion = std::make_unique<Fusion>();
>   FusionGuard fg(fusion.get());
> 
>   auto tv0 = makeSymbolicTensor(2);
>   fusion->addInput(tv0);
>   auto tv1 = set(tv0);
>   auto tv2 = sum(tv1, {1});
>   auto tv3 = broadcast(tv2, {false, true});
>   auto tv4 = add(tv1, tv3);
>   fusion->addOutput(tv4);
> 
>   auto producer = tv1;
>   auto consumer = tv4;
> 
>   const auto c2p_root_map =
>       PairwiseRootDomainMap(producer, consumer)
>           .mapConsumerToProducer(consumer->domain(), producer->domain());
> 
>   // IterDomains in consumer root also in producer root
>   std::unordered_set<Val*> mapped_consumer_roots;
>   for (auto entry : c2p_root_map) {
>     mapped_consumer_roots.emplace(entry.first);
>   }
> 
>   fusion->print();
>   std::cout << ir_utils::toString(mapped_consumer_roots) << std::endl;
> }
> ```
> 
> This gives `iS8{i1}, iS9{i2}`.
> 
> I think "unmapped IDs in consumer but not in producer" here refers to new broadcasting dims, and we are saying that, if we see a leaf ID that completely comes from new broadcasting dims, then we can ignore it. But why don't we want to symmetrically do the same for reductions in the producer? For non-trivial reduction, I think this is because it could not be inlined, but for trivial reduction, it should be safe to just ignore it as well?

The constraint about reductions not being able to get inlined is reflected in `ComputeAtRootDomainMap`. `PairwiseRootDomainMap` does not consider constraints due to inlining but just looks at a pair of a producer and consumer.